server/src/utils/database.ts

```diff
@@ -18,131 +17,54 @@ export function OptionalBetween<T>(from?: T, to?: T) {
   }
 }
 
-export const asVector = (embedding: number[], quote = false) =>
-  quote ? `'[${embedding.join(',')}]'` : `[${embedding.join(',')}]`;
-
-export function searchAssetBuilder(
-  builder: SelectQueryBuilder<AssetEntity>,
-  options: AssetSearchBuilderOptions,
-): SelectQueryBuilder<AssetEntity> {
-  builder.andWhere(
-    _.omitBy(
-      {
-        createdAt: OptionalBetween(options.createdAfter, options.createdBefore),
-        updatedAt: OptionalBetween(options.updatedAfter, options.updatedBefore),
-        deletedAt: OptionalBetween(options.trashedAfter, options.trashedBefore),
-        fileCreatedAt: OptionalBetween(options.takenAfter, options.takenBefore),
-      },
-      _.isUndefined,
-    ),
-  );
-
-  const exifInfo = _.omitBy(_.pick(options, ['city', 'country', 'lensModel', 'make', 'model', 'state']), _.isUndefined);
-  const hasExifQuery = Object.keys(exifInfo).length > 0;
-
-  if (options.withExif && !hasExifQuery) {
-    builder.leftJoinAndSelect(`${builder.alias}.exifInfo`, 'exifInfo');
-  }
-
-  if (hasExifQuery) {
-    if (options.withExif) {
-      builder.leftJoinAndSelect(`${builder.alias}.exifInfo`, 'exifInfo');
-    } else {
-      builder.leftJoin(`${builder.alias}.exifInfo`, 'exifInfo');
-    }
-
-    for (const [key, value] of Object.entries(exifInfo)) {
-      if (value === null) {
-        builder.andWhere(`exifInfo.${key} IS NULL`);
-      } else {
-        builder.andWhere(`exifInfo.${key} = :${key}`, { [key]: value });
-      }
+export const UPSERT_COLUMNS = {} as { [T in keyof DB]: { [K in keyof DB[T]]: RawBuilder<unknown> } };
+/** Any repository that upserts to a table using `mapUpsertColumns` should call this method in its constructor with that table. */
```

Comment:
When an upsert is done, every field that's going to be updated has to be part of the query as `excluded.{field}`. We precompute the `excluded.*` for a table so `mapUpsertColumns` can reference this instead of recomputing it on each upsert.

Reply:
Can you help giving an example to demonstrate the idea?

Reply:
If you have a table with `id`, `path`, `type` and perform an update on `path` and `type`, `mapUpsertColumns` will return an object `{ path: sql\`excluded.${sql.ref('path')\`, type: sql\`excluded.${sql.ref(type)}\` }`. It will exclude `id` from that object because it's in the `conflictKeys`. If you updated only `path`, then `type` would not be in the object either.